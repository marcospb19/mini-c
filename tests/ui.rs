//! This is a test module that creates snapshots, and might do
//! UI testing in the future.
//!
//! This shouldn't be called ui.rs hmmmmmmmm.
//!
//! Snapshot todo list:
//! - [X] Source code.
//! - [X] AST.
//! - [ ] LLVM IR.
//! - [ ] Execution stdout output (UI).
//! - [ ] (maybe) Execution final state (variables, functions).

use fs_err as fs;
use mini_c::test_utils;

macro_rules! snapshot_example_file {
    ($example_path:literal) => {
        let file_contents = &fs::read_to_string($example_path)
            .unwrap_or_else(|err| panic!("Failed to read file {}: {err}.", $example_path));
        // Source code
        insta::assert_display_snapshot!(file_contents);
        // AST
        insta::assert_ron_snapshot!(test_utils::generate_ast(file_contents));
    };
}

#[test]
fn example_break() {
    snapshot_example_file!("examples/break.c");
}
#[test]
fn example_coins() {
    snapshot_example_file!("examples/coins.c");
}
#[test]
fn example_conds() {
    snapshot_example_file!("examples/conds.c");
}
#[test]
fn example_easy() {
    snapshot_example_file!("examples/easy.c");
}
#[test]
fn example_fib() {
    snapshot_example_file!("examples/fib.c");
}
#[test]
fn example_file_1() {
    snapshot_example_file!("examples/file_1.c");
}
#[test]
fn example_file_2() {
    snapshot_example_file!("examples/file_2.c");
}
#[test]
fn example_if_pass() {
    snapshot_example_file!("examples/if-pass.c");
}
#[test]
fn example_operators() {
    snapshot_example_file!("examples/operators.c");
}
#[test]
fn example_queen() {
    snapshot_example_file!("examples/queen.c");
}
#[test]
fn example_var() {
    snapshot_example_file!("examples/var.c");
}
#[test]
fn example_var_loop() {
    snapshot_example_file!("examples/var_loop.c");
}
#[test]
fn example_while() {
    snapshot_example_file!("examples/while.c");
}
