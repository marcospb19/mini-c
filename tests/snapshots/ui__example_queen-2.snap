---
source: tests/ui.rs
expression: test_mini_c(file_contents)
---
Program(
  preamble: true,
  declarations: [
    Variable(VariableDeclaration(Int, [
      Normal("N"),
    ])),
    Variable(VariableDeclaration(Int, [
      Array("a", 20),
      Array("c", 400),
    ])),
    Function(FunctionDeclaration(Void, "markcolumn", [
      Parameter(Int, "x"),
      Parameter(Int, "y"),
      Parameter(Int, "mark"),
    ], Some(Scope([
      VariableDeclaration(Int, [
        Normal("i"),
      ]),
    ], [
      For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, VariableReference(Normal("N"))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([], [
        Expression(Assignment(AssignmentExpression(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("i")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y"))))), Binary(BinaryExpression(VariableReference(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("i")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y")))))), Add, VariableReference(Normal("mark"))))))),
      ]))),
    ])))),
    Function(FunctionDeclaration(Void, "markdiag", [
      Parameter(Int, "x"),
      Parameter(Int, "y"),
      Parameter(Int, "mark"),
    ], Some(Scope([
      VariableDeclaration(Int, [
        Normal("i"),
      ]),
    ], [
      For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, VariableReference(Normal("N"))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([
        VariableDeclaration(Int, [
          Normal("x1"),
        ]),
        VariableDeclaration(Int, [
          Normal("y1"),
        ]),
      ], [
        Expression(Assignment(AssignmentExpression(Normal("x1"), VariableReference(Normal("i"))))),
        Expression(Assignment(AssignmentExpression(Normal("y1"), Binary(BinaryExpression(VariableReference(Normal("y")), Add, Binary(BinaryExpression(VariableReference(Normal("i")), Sub, VariableReference(Normal("x"))))))))),
        If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("y1")), GreaterOrEquals, Value(Int(0)))), And, Binary(BinaryExpression(VariableReference(Normal("y1")), Less, VariableReference(Normal("N")))))), Scope([], [
          Expression(Assignment(AssignmentExpression(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("x1")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y1"))))), Binary(BinaryExpression(VariableReference(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("x1")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y1")))))), Add, VariableReference(Normal("mark"))))))),
        ]), None)),
        Expression(Assignment(AssignmentExpression(Normal("y1"), Binary(BinaryExpression(VariableReference(Normal("y")), Sub, Binary(BinaryExpression(VariableReference(Normal("i")), Sub, VariableReference(Normal("x"))))))))),
        If(IfStatement(Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("y1")), GreaterOrEquals, Value(Int(0)))), And, Binary(BinaryExpression(VariableReference(Normal("y1")), Less, VariableReference(Normal("N")))))), Scope([], [
          Expression(Assignment(AssignmentExpression(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("x1")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y1"))))), Binary(BinaryExpression(VariableReference(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("x1")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("y1")))))), Add, VariableReference(Normal("mark"))))))),
        ]), None)),
      ]))),
    ])))),
    Function(FunctionDeclaration(Void, "search", [
      Parameter(Int, "k"),
    ], Some(Scope([
      VariableDeclaration(Int, [
        Normal("i"),
      ]),
    ], [
      If(IfStatement(Binary(BinaryExpression(VariableReference(Normal("k")), Equals, VariableReference(Normal("N")))), Scope([], [
        For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, VariableReference(Normal("N"))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([], [
          Expression(FunctionCall(FunctionCallExpression("putint", [
            VariableReference(Array("a", VariableReference(Normal("i")))),
          ]))),
        ]))),
        Expression(FunctionCall(FunctionCallExpression("putnewline", []))),
        Return(ReturnStatement(None)),
      ]), None)),
      For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, VariableReference(Normal("N"))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([], [
        If(IfStatement(Binary(BinaryExpression(VariableReference(Array("c", Binary(BinaryExpression(Binary(BinaryExpression(VariableReference(Normal("k")), Mul, VariableReference(Normal("N")))), Add, VariableReference(Normal("i")))))), Equals, Value(Int(0)))), Scope([], [
          Expression(FunctionCall(FunctionCallExpression("markcolumn", [
            VariableReference(Normal("k")),
            VariableReference(Normal("i")),
            Value(Int(1)),
          ]))),
          Expression(FunctionCall(FunctionCallExpression("markdiag", [
            VariableReference(Normal("k")),
            VariableReference(Normal("i")),
            Value(Int(1)),
          ]))),
          Expression(Assignment(AssignmentExpression(Array("a", VariableReference(Normal("k"))), VariableReference(Normal("i"))))),
          Expression(FunctionCall(FunctionCallExpression("search", [
            Binary(BinaryExpression(VariableReference(Normal("k")), Add, Value(Int(1)))),
          ]))),
          Expression(FunctionCall(FunctionCallExpression("markdiag", [
            VariableReference(Normal("k")),
            VariableReference(Normal("i")),
            Unary(UnaryExpression(Negative, Value(Int(1)))),
          ]))),
          Expression(FunctionCall(FunctionCallExpression("markcolumn", [
            VariableReference(Normal("k")),
            VariableReference(Normal("i")),
            Unary(UnaryExpression(Negative, Value(Int(1)))),
          ]))),
        ]), None)),
      ]))),
    ])))),
    Function(FunctionDeclaration(Int, "main", [], Some(Scope([
      VariableDeclaration(Int, [
        Normal("i"),
      ]),
    ], [
      Expression(Assignment(AssignmentExpression(Normal("N"), FunctionCall(FunctionCallExpression("getint", []))))),
      For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, Binary(BinaryExpression(VariableReference(Normal("N")), Mul, VariableReference(Normal("N"))))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([], [
        Expression(Assignment(AssignmentExpression(Array("c", VariableReference(Normal("i"))), Value(Int(0))))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("search", [
        Value(Int(0)),
      ]))),
      Return(ReturnStatement(Some(Value(Int(0))))),
    ])))),
  ],
)
